# Dataset Exploration

## Objective

Understand the UFC dataset before performing any analysis.

## Import Libraries

In [ ]:
import pandas as pd

## Load Dataset

In [ ]:
# save filepath to variable for easier access
ufc_file_path = '../data/raw/ufc-master.csv'

# read the data and store data in DataFrame titled ufc_data
ufc_data = pd.read_csv(ufc_file_path)

## Dataset Overview

In [ ]:
# print summary of data
ufc_data.describe()

,R_odds,B_odds,R_ev,B_ev,no_of_rounds,B_current_lose_streak,B_current_win_streak,B_draw,B_avg_SIG_STR_landed,B_avg_SIG_STR_pct,...,B_Flyweight_rank,B_Pound-for-Pound_rank,finish_round,total_fight_time_secs,r_dec_odds,b_dec_odds,r_sub_odds,b_sub_odds,r_ko_odds,b_ko_odds
count,6937.000000,6938.000000,6937.000000,6938.000000,7177.000000,7177.000000,7177.000000,7177.000000,6247.000000,6412.000000,...,143.000000,81.000000,6555.000000,6555.000000,6062.000000,6032.000000,5813.000000,5789.000000,5815.000000,5788.000000
mean,-114.380280,56.003459,98.775801,165.136205,3.187126,0.502996,1.002229,0.025916,18.211118,0.456023,...,8.440559,9.370370,2.423036,657.684821,315.757176,431.808687,995.132118,1211.647608,596.026827,729.980477
std,285.370533,260.799641,90.112714,140.153769,0.579962,0.797343,1.465126,0.164074,19.825099,0.108546,...,4.284960,4.357305,1.009258,360.670777,260.648779,337.468752,816.926221,903.365182,638.926484,699.021015
min,-2100.000000,-2200.000000,4.761900,4.545500,3.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,1.000000,1.000000,1.000000,5.000000,-440.000000,-200.000000,-370.000000,-1250.000000,-900.000000,-400.000000
25%,-255.000000,-150.000000,39.215700,66.666700,3.000000,0.000000,0.000000,0.000000,3.745000,0.400000,...,5.000000,5.000000,1.000000,299.000000,170.000000,220.000000,450.000000,600.000000,240.000000,320.000000
50%,-148.000000,130.000000,67.567600,130.000000,3.000000,0.000000,0.000000,0.000000,6.120000,0.460000,...,8.000000,10.000000,3.000000,900.000000,260.000000,350.000000,800.000000,1000.000000,450.000000,550.000000
75%,130.000000,215.000000,130.000000,215.000000,3.000000,1.000000,1.000000,0.000000,30.433050,0.520000,...,12.000000,14.000000,3.000000,900.000000,420.000000,550.000000,1300.000000,1550.000000,762.500000,900.000000
max,1100.000000,1300.000000,1100.000000,1300.000000,5.000000,6.000000,15.000000,2.000000,154.000000,1.000000,...,15.000000,15.000000,5.000000,1500.000000,2400.000000,3500.000000,10000.000000,12500.000000,5000.000000,6600.000000


In [ ]:
ufc_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 7177 entries, 0 to 7176
Columns: 118 entries, R_fighter to b_ko_odds
dtypes: bool(1), float64(60), int64(43), str(14)
memory usage: 6.4 MB


## What does one row represent?

In [ ]:
ufc_data.columns.to_list()

['R_fighter',
 'B_fighter',
 'R_odds',
 'B_odds',
 'R_ev',
 'B_ev',
 'date',
 'location',
 'country',
 'Winner',
 'title_bout',
 'weight_class',
 'gender',
 'no_of_rounds',
 'B_current_lose_streak',
 'B_current_win_streak',
 'B_draw',
 'B_avg_SIG_STR_landed',
 'B_avg_SIG_STR_pct',
 'B_avg_SUB_ATT',
 'B_avg_TD_landed',
 'B_avg_TD_pct',
 'B_longest_win_streak',
 'B_losses',
 'B_total_rounds_fought',
 'B_total_title_bouts',
 'B_win_by_Decision_Majority',
 'B_win_by_Decision_Split',
 'B_win_by_Decision_Unanimous',
 'B_win_by_KO/TKO',
 'B_win_by_Submission',
 'B_win_by_TKO_Doctor_Stoppage',
 'B_wins',
 'B_Stance',
 'B_Height_cms',
 'B_Reach_cms',
 'B_Weight_lbs',
 'R_current_lose_streak',
 'R_current_win_streak',
 'R_draw',
 'R_avg_SIG_STR_landed',
 'R_avg_SIG_STR_pct',
 'R_avg_SUB_ATT',
 'R_avg_TD_landed',
 'R_avg_TD_pct',
 'R_longest_win_streak',
 'R_losses',
 'R_total_rounds_fought',
 'R_total_title_bouts',
 'R_win_by_Decision_Majority',
 'R_win_by_Decision_Split',
 'R_win_by_Decision_Un

### Findings

- Calling `dropna()` removes every row.
- Missing values are largely structural.
- Women's ranking columns are mostly empty because most fights aren't in those divisions.

In [ ]:
ufc_data.shape

(7177, 118)

In [ ]:
ufc_data.isna().sum().sort_values(ascending=False)

B_Women's Featherweight_rank    7176
R_Women's Featherweight_rank    7168
B_Pound-for-Pound_rank          7096
B_Women's Flyweight_rank        7092
R_Women's Flyweight_rank        7066
                                ... 
win_dif                            0
loss_dif                           0
total_round_dif                    0
total_title_bout_dif               0
ko_dif                             0
Length: 118, dtype: int64

In [ ]:
print(ufc_data[["Winner"]])

     Winner
0      Blue
1       Red
2       Red
3      Blue
4      Blue
...     ...
7172   Blue
7173    Red
7174    Red
7175    Red
7176   Blue

[7177 rows x 1 columns]


## Find every unique weight class.

In [ ]:
ufc_data["weight_class"].unique()

<StringArray>
[         'Middleweight',     'Women's Flyweight',          'Welterweight',
         'Featherweight',           'Lightweight',           'Heavyweight',
     'Light Heavyweight',          'Bantamweight',   'Women's Strawweight',
             'Flyweight',  'Women's Bantamweight',          'Catch Weight',
 'Women's Featherweight']
Length: 13, dtype: str

## Which division has most fights?

In [ ]:
ufc_data['weight_class'].value_counts()

weight_class
Lightweight              1160
Welterweight             1097
Middleweight              873
Featherweight             833
Bantamweight              753
Light Heavyweight         551
Heavyweight               541
Flyweight                 404
Women's Strawweight       359
Women's Flyweight         268
Women's Bantamweight      237
Catch Weight               71
Women's Featherweight      30
Name: count, dtype: int64

In [ ]:
ufc_data['weight_class'].value_counts(normalize=True)

weight_class
Lightweight              0.161627
Welterweight             0.152849
Middleweight             0.121639
Featherweight            0.116065
Bantamweight             0.104918
Light Heavyweight        0.076773
Heavyweight              0.075380
Flyweight                0.056291
Women's Strawweight      0.050021
Women's Flyweight        0.037342
Women's Bantamweight     0.033022
Catch Weight             0.009893
Women's Featherweight    0.004180
Name: proportion, dtype: float64

### Every time we investigate a categorical column, ask these three questions:

- What categories exist?
    unique()
- How many of each are there?
    value_counts()
- What proportion does each category represent?
    value_counts(normalize=True)

You'll use this pattern dozens of times in this project.

## Target Variable

In [ ]:
y = ufc_data['Winner']

### Is the dataset balanced about the Winning variable?

In [ ]:
ufc_data['Winner'].value_counts(normalize=True)

Winner
Red           0.577539
Blue          0.421346
Draw          0.000697
No Contest    0.000418
Name: proportion, dtype: float64

### Findings

Winner distribution:

- Red: 57.75%
- Blue: 42.13%
- Draw: very rare
- No Contest: very rare

A baseline model predicting "Red" would achieve about 58% accuracy.

## Datatypes

In [ ]:
ufc_data.dtypes

R_fighter         str
B_fighter         str
R_odds        float64
B_odds        float64
R_ev          float64
               ...   
b_dec_odds    float64
r_sub_odds    float64
b_sub_odds    float64
r_ko_odds     float64
b_ko_odds     float64
Length: 118, dtype: object

In [ ]:
ufc_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 7177 entries, 0 to 7176
Columns: 118 entries, R_fighter to b_ko_odds
dtypes: bool(1), float64(60), int64(43), str(14)
memory usage: 6.4 MB


## Checking for duplicates

In [ ]:
ufc_data.duplicated().sum()

np.int64(0)

No duplicates present, would've used pd.DataFrame.drop_duplicates if needed.

## Key Takeaways

- Each row represents one UFC fight between a Red and Blue fighter.
- The dataset contains 7,177 fights and 118 features.
- Missing values are often structural (e.g., ranking columns for irrelevant weight classes), so `dropna()` is not an appropriate first cleaning step.
- The target variable (`Winner`) is reasonably balanced (~58% Red, ~42% Blue), with very few Draws and No Contests.
- The dataset contains no duplicate rows.
- Several columns (e.g., `finish_round`, `finish_details`, `total_fight_time_secs`) would cause data leakage if used as features.


## Next Question

Does having a reach advantage increase the probability of winning?